include/unifex/finally.hpp

```diff
@@ -0,0 +1,679 @@
+            static_cast<Args&&>(args)...);
+      }
+
+      template <typename Func>
+      friend void tag_invoke(
+          tag_t<visit_continuations>,
+          const finally_error_receiver& r,
+          Func&& func) {
+        std::invoke(func, static_cast<const Receiver&>(r.op_->receiver_));
+      }
+
+      friend unstoppable_token tag_invoke(
+          tag_t<get_stop_token>, const finally_error_receiver&) noexcept {
+        return {};
+      }
+
+    private:
+      operation_type* op_;
+    };
+
+    template <
+        typename SourceSender,
+        typename CompletionSender,
+        typename Receiver>
+    class finally_done_receiver {
+      using operation_type =
+          finally_operation<SourceSender, CompletionSender, Receiver>;
+
+    public:
+      explicit finally_done_receiver(operation_type* op) noexcept : op_(op) {}
+
+      finally_done_receiver(finally_done_receiver&& other) noexcept
+        : op_(std::exchange(other.op_, nullptr)) {}
+
+      void set_value() && noexcept {
+        auto* op = op_;
+        op->completionDoneOp_.destruct();
+        unifex::set_done(static_cast<Receiver&&>(op->receiver_));
+      }
+
+      template <
+          typename Error,
+          std::enable_if_t<
+              std::is_invocable_v<decltype(unifex::set_error), Receiver, Error>,
+              int> = 0>
+      void set_error(Error&& error) && noexcept {
+        auto* op = op_;
+        op->completionDoneOp_.destruct();
+        unifex::set_error(
+            static_cast<Receiver&&>(op->receiver_),
+            static_cast<Error&&>(error));
+      }
+
+      void set_done() && noexcept {
+          auto* op = op_;
+        op->completionDoneOp_.destruct();
+        unifex::set_done(static_cast<Receiver&&>(op->receiver_));
+      }
+
+      template <
+          typename CPO,
+          typename... Args,
+          std::enable_if_t<!is_receiver_cpo_v<CPO>, int> = 0>
+      friend auto tag_invoke(
+          CPO cpo,
+          const finally_done_receiver& r,
+          Args&&... args) noexcept(std::
+                                       is_nothrow_invocable_v<
+                                           CPO,
+                                           const Receiver&,
+                                           Args...>)
+          -> std::invoke_result_t<CPO, const Receiver&, Args...> {
+        return static_cast<CPO&&>(cpo)(
+            static_cast<const Receiver&>(r.op_->receiver_),
+            static_cast<Args&&>(args)...);
+      }
+
+      template <typename Func>
+      friend void tag_invoke(
+          tag_t<visit_continuations>,
+          const finally_done_receiver& r,
+          Func&& func) {
+        std::invoke(func, static_cast<const Receiver&>(r.op_->receiver_));
+      }
+
+      friend unstoppable_token
+      tag_invoke(tag_t<get_stop_token>, const finally_done_receiver&) noexcept {
+        return {};
+      }
+
+    private:
+      operation_type* op_;
+    };
```

Comment:
All the `finally_(value|error|done)_receiver` class templates are very, very similar. Would be nice if there was some way to avoid spamming out nearly identical code three times.

Reply:
We could probably factor out some of the common bits (like forwarding through of CPOs) to a CRTP base-class to reduce some of the duplication. I think the logic on the set_value/error/done methods probably needs to stay separate though.

Reply:
I am used to ding this as a template type that implements the common parts and stores a function that is invoked as the body of the method being parameterized.